invenio_files_rest/helpers.py

```diff
@@ -147,7 +146,7 @@ def send_stream(
         try:
             filenames = {"filename": filename.encode("latin-1")}
         except UnicodeEncodeError:
-            filenames = {"filename*": "UTF-8''%s" % url_quote(filename)}
+            filenames = {"filename*": "UTF-8''%s" % quote(filename)}
```

Comment:
```suggestion
            filenames = {"filename*": "UTF-8''%s" % quote(filename, safe="!#$&+-.^_`|~")}
```

I see in Zenodo after some post-migration fixes we explicitly [pass a `safe=...` parameter](https://github.com/zenodo/zenodo-rdm/blob/8b81afaeba88ba00230bfcf189f6b62b9ef8adb0/site/zenodo_rdm/files.py#L125). I suspect this list comes from [this werkzeug util](https://github.com/pallets/werkzeug/blob/5add63c955131fd73531d7369f16b2f1b4e342d4/src/werkzeug/utils.py#L462C49-L462C61).

Does it make sense to reuse here, or have a look at the `werkzeug` code to verify we're roughly doing the same handling? Unfortunately, we're pinned to a lower version of werkzeug, so we can just upgrade and reuse the upstream function.

Reply:
looking into it some more, `filename*` is an *extended parameter* for the header-field *Content-Disposition*
*extended parameter*s for header-fields are defined in [RFC 8187](https://www.rfc-editor.org/rfc/rfc8187)
in its [section 3.2.1](https://www.rfc-editor.org/rfc/rfc8187#section-3.2.1), the exact characters used by `werkzeug` (`"!#$&+-.^_`|~"`) are defined as safe to use without percent-encoding as values for extended fields

so adding `safe="!#$&+-.^_`|~"` makes sense imo

Reply:
I think if at all we should use `flask.send_file` instead of `werkzeug.utils.send_file`

If I understand correctly, there's a reason that `invenio`'s function is called `send_stream` instead of `send_file` though:
we're working with an abstract file-system provided by `pyfs`, so the only thing we're guaranteed to have is a *stream*, not necessary a *file* (e.g. could be entry in a zip-file, abstracted by `pyfs` to behave like a filesystem)

now, streams can be passed to `flask.send_file` and work just fine in my local testing
*but*, we'd lose *Content-Length* and *Last-Modified* headers
(`flask.send_file` tries to get those headers from the passed file, but leaves them out when passed a stream)
(`invenio_files_rest.helpers.send_stream` has additional parameters for these headers instead)

not sure how easy those could be added if we were to use `flask.send_file`...

Reply:
Comparing our current code with `werkzeug 3.0`, there's some differences wrt *Content-Disposition* header:
- `werkzeug` adds `filename=...` and `filename*=..` to *Content-Disposition* whether it's `inline` or `attachment`, whereas we just add it in the `attachment` case
- `werkzeug`'s `filename` is `ascii` whereas ours is `latin-1`

*Content-Disposition* header is defined in [RFC 62266](https://www.rfc-editor.org/rfc/rfc6266)
in [its *Appendix D*](https://www.rfc-editor.org/rfc/rfc6266#appendix-D), it lists some advice for this header
reading that, I think `werkzeug` is the more correct one in both of the cases I listed above...

there might be further differences between `invenio_files_rest.helpers.stream_file` and `flask.send_file` for other headers...